mlrun/track/tracker_manager.py

```
@@ -0,0 +1,92 @@
+
+from mlrun.execution import MLClientCtx
+from mlrun.utils import logger
+
+from .tracker import Tracker
+
+
+class TrackerManager:
+    """
+    A class for handling multiple `Tracker` instances during a run. The manager is a singleton class and should be
+    retrieved using the `mlrun.track.get_trackers_manager` function.
+    """
+
+    def __init__(self, _stale: bool = False):
+        """
+        Initialize a new empty tracker manager.
+
+        :param _stale: An inner attribute to init a trackers manager in a specific staleness state.
+        """
+        self._trackers: List[Type[Tracker]] = []
+        self._stale = _stale
+
+    def add_tracker(self, tracker: Type[Tracker]):
+        """
+        Adds a Tracker to trackers list.
+
+        :param tracker: The tracker class to add
+        """
+        self._trackers.append(tracker)
+        logger.debug("Added tracker", tracker=tracker)
+
+    def pre_run(self, context: MLClientCtx):
+        """
+        Goes over all the managed trackers and calls their `pre_run` method.
+
+        :param context: Active mlrun context.
+        """
+        for tracker in self._trackers:
+            tracker.pre_run(context=context)
+
+    def post_run(self, context: Union[MLClientCtx, dict]) -> Union[MLClientCtx, dict]:
+        """
+        Goes over all the managed trackers and calls their `post_run` method.
+
+        :param context: Active mlrun context.
+
+        :return: The context updated with the trackers products.
+        """
+        # Check if the context received is a dict to initialize it as an `MLClientCtx` object:
+
+        is_context_dict = False
+        if isinstance(context, dict):
+            is_context_dict = True
+            context = MLClientCtx.from_dict(
+                context, include_status=True, store_run=False
+            )
+
+        # Go over the trackers and call the `post_run` method:
+        for tracker in self._trackers:
+            tracker.post_run(context)
```

same here about fail fast or not

---

added try-catch